File: mobile/src/main/java/uk/co/sentinelweb/bitwatcher/common/ui/picker/PickerView.java
package uk.co.sentinelweb.bitwatcher.common.ui.picker;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Rect;
import android.support.annotation.DimenRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.Adapter;
import android.support.v7.widget.RecyclerView.ItemDecoration;
import android.support.v7.widget.RecyclerView.State;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnLayoutChangeListener;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import uk.co.sentinelweb.bitwatcher.R;

public class PickerView<T> implements PickerContract.View<T> {

    private static final int SIZE_ANIM_DURATION = 100;
    private final RecyclerView recyclerView;

    private final Context context;

    private int previousPosition;

    private PickerContract.Presenter<T> presenter;

    public PickerView(@NonNull View view) {
        this.context = view.getContext();
        recyclerView = (RecyclerView) view;
        setupRecyclerView();
    }

    @Override
    public void setPresenter(@NonNull PickerContract.Presenter<T> presenter) {
        this.presenter = presenter;
    }

    @Override
    public void setData(@NonNull List<PickerItemModel<T>> ages) {
        recyclerView.setAdapter(new ItemViewHolderAdapter(ages));
    }

    @Override
    public void setInitialPosition(int position) {
        previousPosition = position;
        if (recyclerView.getMeasuredHeight() > 0) {
            initialSelectionOfPosition(position);
        } else {
            recyclerView.addOnLayoutChangeListener(new SelectPositionAfterLayoutListener(position));
        }
    }

    private void initialSelectionOfPosition(int position) {
        goToPosition(position);
        View selectedView = PickerView.this.recyclerView.getLayoutManager()
                .findViewByPosition(position);
        setItemSelected(selectedView);
        recyclerView.addOnScrollListener(new SpinnerOnScrollListener());
    }

    private void goToPosition(int position) {
        int itemsHeight = recyclerView.computeVerticalScrollRange() - recyclerView.getMeasuredHeight();
        int ypos = (int) ((position / (float) recyclerView.getAdapter().getItemCount()) * itemsHeight);
        recyclerView.scrollBy(0, ypos - recyclerView.getScrollY());
    }

    private void setupRecyclerView() {
        final LinearLayoutManager layout = new LinearLayoutManager(context);
        recyclerView.setLayoutManager(layout);

        // adds item decoration so recycler view can scroll first and last item to the middle for selection
        recyclerView.addItemDecoration(new SpinnerEndOffsetToMiddleItemDecoration());
    }

    private void changeTextSize(@NonNull TextView tv, @DimenRes int newSizeResource) {
        int newSize = context.getResources().getDimensionPixelSize(newSizeResource);
        ValueAnimator animator = ValueAnimator.ofFloat(tv.getTextSize(), newSize);
        animator.setDuration(SIZE_ANIM_DURATION);
        animator.addUpdateListener(new TextSizeAnimationListener(tv));
        animator.start();
    }

    private void setItemSelected(View selectedView) {
        if (selectedView != null) {
            selectedView.setSelected(true);
            changeTextSize((TextView) selectedView, R.dimen.text_size_large);
        }
    }

    private class ItemViewHolder extends RecyclerView.ViewHolder {

        final TextView view;

        ItemViewHolder(@NonNull TextView view) {
            super(view);
            this.view = view;
        }
    }

    private class ItemViewHolderAdapter extends Adapter<ItemViewHolder> {

        private final List<PickerItemModel<T>> itemModels;

        public ItemViewHolderAdapter(@NonNull List<PickerItemModel<T>> itemModels) {
            this.itemModels = itemModels;
        }

        @Override
        public ItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new ItemViewHolder((TextView) LayoutInflater
                    .from(context).inflate(R.layout.view_picker_item, parent, false));
        }

        @Override
        public void onBindViewHolder(ItemViewHolder holder, int position) {
            holder.view.setSelected(false);
            holder.view.setTextSize(TypedValue.COMPLEX_UNIT_PX,
                    context.getResources().getDimensionPixelSize(R.dimen.text_size_medium));
            holder.view.setText(itemModels.get(position).display);
        }

        @Override
        public int getItemCount() {
            return itemModels.size();
        }
    }

    private class SpinnerOnScrollListener extends RecyclerView.OnScrollListener {

        @Override
        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            super.onScrolled(recyclerView, dx, dy);

            View centerView = PickerView.this.recyclerView.findChildViewUnder(
                    PickerView.this.recyclerView.getMeasuredWidth() / 2,
                    PickerView.this.recyclerView.getMeasuredHeight() / 2);
            int centerPos = PickerView.this.recyclerView.getChildAdapterPosition(centerView);

            if (centerPos > -1 && previousPosition != centerPos) {
                View previouslySelectedView = PickerView.this.recyclerView.getLayoutManager()
                        .findViewByPosition(previousPosition);
                if (previouslySelectedView != null) {
                    previouslySelectedView.setSelected(false);
                    changeTextSize((TextView) previouslySelectedView, R.dimen.text_size_medium);
                }

                setItemSelected(centerView);

                previousPosition = centerPos;
                presenter.updatePositionFromScroll(previousPosition); // is current at the moment
            }
        }
    }

    private class SpinnerEndOffsetToMiddleItemDecoration extends ItemDecoration {

        @Override public void getItemOffsets(Rect outRect, View view, RecyclerView parent,
                                             State state) {
            int viewPosition = recyclerView.getChildAdapterPosition(view);
            if (viewPosition == 0) {
                outRect.top = (recyclerView.getHeight() - view.getHeight()) / 2;
            } else if (viewPosition == parent.getAdapter().getItemCount() - 1) {
                outRect.bottom = (recyclerView.getHeight() - view.getHeight()) / 2;
            } else {
                super.getItemOffsets(outRect, view, parent, state);
            }
        }
    }

    private class SelectPositionAfterLayoutListener implements OnLayoutChangeListener {

        private final int position;

        public SelectPositionAfterLayoutListener(int position) {
            this.position = position;
        }

        @Override public void onLayoutChange(View v, int left, int top, int right,
                                             int bottom,
                                             int oldLeft,
                                             int oldTop, int oldRight, int oldBottom) {
            recyclerView.removeOnLayoutChangeListener(this);
            initialSelectionOfPosition(position);
        }
    }

    private static class TextSizeAnimationListener implements ValueAnimator.AnimatorUpdateListener {

        private final TextView tv;

        public TextSizeAnimationListener(@NonNull TextView tv) {
            this.tv = tv;
        }

        @Override
        public void onAnimationUpdate(ValueAnimator valueAnimator) {
            float animatedValue = (float) valueAnimator.getAnimatedValue();
            tv.setTextSize(TypedValue.COMPLEX_UNIT_PX, animatedValue);
        }
    }
}

File: mobile/src/main/java/uk/co/sentinelweb/bitwatcher/common/ui/picker/PickerView.kt
package uk.co.sentinelweb.bitwatcher.common.ui.picker

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Rect
import android.support.annotation.DimenRes
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.Adapter
import android.support.v7.widget.RecyclerView.ItemDecoration
import android.support.v7.widget.RecyclerView.State
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLayoutChangeListener
import android.view.ViewGroup
import android.widget.TextView

import uk.co.sentinelweb.bitwatcher.R

class PickerView<T>(view: View) : PickerContract.View<T> {
    private val recyclerView: RecyclerView

    private val context: Context

    private var previousPosition: Int = 0

    private var presenter: PickerContract.Presenter<T>? = null

    init {
        this.context = view.context
        recyclerView = view as RecyclerView
        setupRecyclerView()
    }

    override fun setPresenter(presenter: PickerContract.Presenter<T>) {
        this.presenter = presenter
    }

    override fun setData(ages: List<PickerItemModel<T>>) {
        recyclerView.adapter = ItemViewHolderAdapter(ages)
    }

    override fun setInitialPosition(position: Int) {
        previousPosition = position
        if (recyclerView.measuredHeight > 0) {
            initialSelectionOfPosition(position)
        } else {
            recyclerView.addOnLayoutChangeListener(SelectPositionAfterLayoutListener(position))
        }
    }

    private fun initialSelectionOfPosition(position: Int) {
        goToPosition(position)
        val selectedView = this@PickerView.recyclerView.layoutManager
                .findViewByPosition(position)
        setItemSelected(selectedView)
        recyclerView.addOnScrollListener(SpinnerOnScrollListener())
    }

    private fun goToPosition(position: Int) {
        val itemsHeight = recyclerView.computeVerticalScrollRange() - recyclerView.measuredHeight
        val ypos = (position / recyclerView.adapter.itemCount.toFloat() * itemsHeight).toInt()
        recyclerView.scrollBy(0, ypos - recyclerView.scrollY)
    }

    private fun setupRecyclerView() {
        val layout = LinearLayoutManager(context)
        recyclerView.layoutManager = layout

        // adds item decoration so recycler view can scroll first and last item to the middle for selection
        recyclerView.addItemDecoration(SpinnerEndOffsetToMiddleItemDecoration())
    }

    private fun changeTextSize(tv: TextView, @DimenRes newSizeResource: Int) {
        val newSize = context.resources.getDimensionPixelSize(newSizeResource)
        val animator = ValueAnimator.ofFloat(tv.textSize, newSize.toFloat())
        animator.duration = SIZE_ANIM_DURATION.toLong()
        animator.addUpdateListener(TextSizeAnimationListener(tv))
        animator.start()
    }

    private fun setItemSelected(selectedView: View?) {
        if (selectedView != null) {
            selectedView.isSelected = true
            changeTextSize((selectedView as TextView?)!!, R.dimen.text_size_large)
        }
    }

    private inner class ItemViewHolder internal constructor(internal val view: TextView) : RecyclerView.ViewHolder(view)

    private inner class ItemViewHolderAdapter(private val itemModels: List<PickerItemModel<T>>) : Adapter<ItemViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
            return ItemViewHolder(LayoutInflater
                    .from(context).inflate(R.layout.view_picker_item, parent, false) as TextView)
        }

        override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
            holder.view.isSelected = false
            holder.view.setTextSize(TypedValue.COMPLEX_UNIT_PX,
                    context.resources.getDimensionPixelSize(R.dimen.text_size_medium).toFloat())
            holder.view.text = itemModels[position].display
        }

        override fun getItemCount(): Int {
            return itemModels.size
        }
    }

    private inner class SpinnerOnScrollListener : RecyclerView.OnScrollListener() {

        override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)

            val centerView = this@PickerView.recyclerView.findChildViewUnder(
                    (this@PickerView.recyclerView.measuredWidth / 2).toFloat(),
                    (this@PickerView.recyclerView.measuredHeight / 2).toFloat())
            val centerPos = this@PickerView.recyclerView.getChildAdapterPosition(centerView)

            if (centerPos > -1 && previousPosition != centerPos) {
                val previouslySelectedView = this@PickerView.recyclerView.layoutManager
                        .findViewByPosition(previousPosition)
                if (previouslySelectedView != null) {
                    previouslySelectedView.isSelected = false
                    changeTextSize(previouslySelectedView as TextView, R.dimen.text_size_medium)
                }

                setItemSelected(centerView)

                previousPosition = centerPos
                presenter!!.updatePositionFromScroll(previousPosition) // is current at the moment
            }
        }
    }

    private inner class SpinnerEndOffsetToMiddleItemDecoration : ItemDecoration() {

        override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView,
                                    state: State?) {
            val viewPosition = recyclerView.getChildAdapterPosition(view)
            if (viewPosition == 0) {
                outRect.top = (recyclerView.height - view.height) / 2
            } else if (viewPosition == parent.adapter.itemCount - 1) {
                outRect.bottom = (recyclerView.height - view.height) / 2
            } else {
                super.getItemOffsets(outRect, view, parent, state)
            }
        }
    }

    private inner class SelectPositionAfterLayoutListener(private val position: Int) : OnLayoutChangeListener {

        override fun onLayoutChange(v: View, left: Int, top: Int, right: Int,
                                    bottom: Int,
                                    oldLeft: Int,
                                    oldTop: Int, oldRight: Int, oldBottom: Int) {
            recyclerView.removeOnLayoutChangeListener(this)
            initialSelectionOfPosition(position)
        }
    }

    private class TextSizeAnimationListener(private val tv: TextView) : ValueAnimator.AnimatorUpdateListener {

        override fun onAnimationUpdate(valueAnimator: ValueAnimator) {
            val animatedValue = valueAnimator.animatedValue as Float
            tv.setTextSize(TypedValue.COMPLEX_UNIT_PX, animatedValue)
        }
    }

    companion object {

        private val SIZE_ANIM_DURATION = 100
    }
}
